config/config.go

```diff
@@ -31,37 +33,80 @@ var (
 	ErrEmpty       = errors.New("not a config (empty)")
 	ErrScript      = errors.New("not a config (found coreos-cloudinit script)")
 	ErrDeprecated  = errors.New("config format deprecated")
+	ErrInvalid     = errors.New("config is not valid")
 )
 
-func Parse(rawConfig []byte) (types.Config, error) {
+// ParseReport parses the raw config into a types.Config struct and generates a report of any
+// errors, warnings, info and deprecations it encountered
+func Parse(rawConfig []byte) (types.Config, report.Report, error) {
+	if isEmpty(rawConfig) {
+		return types.Config{}, report.Report{}, ErrEmpty
+	} else if isCloudConfig(rawConfig) {
+		return types.Config{}, report.Report{}, ErrCloudConfig
+	} else if isScript(rawConfig) {
+		return types.Config{}, report.Report{}, ErrScript
+	}
+
 	switch majorVersion(rawConfig) {
 	case 1:
 		config, err := ParseFromV1(rawConfig)
 		if err != nil {
-			return types.Config{}, err
+			return types.Config{}, report.ReportFromError(err, report.EntryError), err
 		}
 
-		return config, ErrDeprecated
+		return config, report.ReportFromError(ErrDeprecated, report.EntryDeprecated), nil
 	default:
 		return ParseFromLatest(rawConfig)
 	}
 }
 
-func ParseFromLatest(rawConfig []byte) (config types.Config, err error) {
+func ParseFromLatest(rawConfig []byte) (config types.Config, r report.Report, err error) {
+	// These errors are fatal and the config should not be further validated
 	if err = json.Unmarshal(rawConfig, &config); err == nil {
 		err = config.Ignition.Version.AssertValid()
-	} else if isEmpty(rawConfig) {
-		err = ErrEmpty
-	} else if isCloudConfig(rawConfig) {
-		err = ErrCloudConfig
-	} else if isScript(rawConfig) {
-		err = ErrScript
 	}
+
+	// Handle json syntax and type errors first, since they are fatal but have offset info
 	if serr, ok := err.(*json.SyntaxError); ok {
 		line, col, highlight := errorutil.HighlightBytePosition(bytes.NewReader(rawConfig), serr.Offset)
-		err = fmt.Errorf("error at line %d, column %d\n%s%v", line, col, highlight, err)
+		r.Add(report.Entry{
+			Kind:      report.EntryError,
+			Message:   serr.Error(),
+			Line:      line,
+			Column:    col,
+			Highlight: highlight,
+		})
+		err = ErrInvalid
+		return
 	}
 
+	if terr, ok := err.(*json.UnmarshalTypeError); ok {
+		line, col, highlight := errorutil.HighlightBytePosition(bytes.NewReader(rawConfig), terr.Offset)
+		r.Add(report.Entry{
+			Kind:      report.EntryError,
+			Message:   terr.Error(),
+			Line:      line,
+			Column:    col,
+			Highlight: highlight,
+		})
+		err = ErrInvalid
+		return
+	}
+
+	// Handle other fatal errors (i.e. invalid version)
+	if err != nil {
+		r = report.ReportFromError(err, report.EntryError)
+		return
+	}
+
+	// Unmarshal again to a json.Node to get offset information for building a report
+	var ast json.Node
+	json.Unmarshal(rawConfig, &ast)
```

Comment:
This should never error. We have a fallback validation function which doesn't report line numbers. should we fall back to using that, and log a warning like "Ignition could not unmarshal your config for reporting line numbers. This is a bug. Please report it" ? Or should we just fail hard?